experimental/devices/mcp9808/mcp9808.go

```diff
@@ -0,0 +1,345 @@
+		return errWritingCritAlert
+	}
+	d.critical = t
+	return nil
+}
+
+func (d *Dev) setUpperAlert(t physic.Temperature) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if t == d.upper {
+		return nil
+	}
+	upper, err := alertTemperatureToBits(t)
+	if err != nil {
+		return err
+	}
+	if err := d.m.WriteUint16(upperAlert, upper); err != nil {
+		return errWritingUpperAlert
+	}
+	d.upper = t
+	return nil
+}
+
+func (d *Dev) setLowerAlert(t physic.Temperature) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if t == d.lower {
+		return nil
+	}
+	lower, err := alertTemperatureToBits(t)
+	if err != nil {
+		return err
+	}
+	if err := d.m.WriteUint16(lowerAlert, lower); err != nil {
+		return errWritingLowerAlert
+	}
+	d.lower = t
+	return nil
+}
+
+type Alert struct {
+	AlertMode  string
+	AlertLevel physic.Temperature
+}
+
+const (
+	// Register addresses.
+	configuration    byte = 0x01
+	upperAlert       byte = 0x02
+	lowerAlert       byte = 0x03
+	critAlert        byte = 0x04
+	temperature      byte = 0x05
+	manifactureID    byte = 0x06
+	deviceID         byte = 0x07
+	resolutionConfig byte = 0x08
+)
+
+var (
+	errReadTemperature      = errors.New("failed to read ambient temperature")
+	errReadCriticalAlert    = errors.New("failed to read critical temperature")
+	errReadUpperAlert       = errors.New("failed to read upper temperature")
+	errReadLowerAlert       = errors.New("failed to read lower temperature")
+	errAddressOutOfRange    = errors.New("i2c address out of range")
+	errInvalidResolution    = errors.New("invalid resolution")
+	errWritingResolution    = errors.New("failed to write resolution configuration")
+	errWritingConfiguration = errors.New("failed to write configuration")
+	errWritingCritAlert     = errors.New("failed to write critical alert configuration")
+	errWritingUpperAlert    = errors.New("failed to write upper alert configuration")
+	errWritingLowerAlert    = errors.New("failed to write lower alert configuration")
+	errAlertOutOfRange      = errors.New("alert seeting exceeds operating conditions")
+	errAlertInvalid         = errors.New("invalid alert temperature configuration")
+)
+
+func alertBitsToTemperature(b uint16) physic.Temperature {
+	b = (b >> 2) & 0x07FF
+	t := physic.Temperature(b&0x03FF) * 250 * physic.MilliKelvin
+	if b&0x400 > 0 {
+		t = -t
+	}
+	t += physic.ZeroCelsius
+	return t
+}
+
+func alertTemperatureToBits(t physic.Temperature) (uint16, error) {
+	const (
+		maxAlert = 125*physic.Kelvin + physic.ZeroCelsius
+		minAlert = -40*physic.Kelvin + physic.ZeroCelsius
+	)
+	if t >= maxAlert || t <= minAlert {
+		return 0, errAlertOutOfRange
+	}
+	t -= physic.ZeroCelsius
+	t /= 250 * physic.MilliKelvin
+
+	var bits uint16
+	if t < 0 {
+		t = -t
+		bits |= 0x400
+	}
+	bits |= uint16(t)
+	bits = bits << 2
+	return bits, nil
+}
+
+type resolution uint8
+
+const (
+	Maximum resolution = 0
+	Low     resolution = 1
+	Medium  resolution = 2
+	High    resolution = 3
+)
+
+var _ conn.Resource = &Dev{}
```

Comment:
```
var _ physic.SenseEnv = &Dev{}
```

Reply:
I had not thought of implementing SenseEnv. I'm not sure about this.